queue/ring.go

```diff
@@ -128,9 +129,24 @@ L:
 // to the queue or Dispose is called on the queue.  An error will be returned
 // if the queue is disposed.
 func (rb *RingBuffer) Get() (interface{}, error) {
-	var n *node
-	pos := atomic.LoadUint64(&rb.dequeue)
-	i := 0
+	return rb.Poll(0)
+}
+
+// Poll will return the next item in the queue.  This call will block
+// if the queue is empty.  This call will unblock when an item is added
+// to the queue, Dispose is called on the queue, or the timeout is reached. An
+// error will be returned if the queue is disposed or a timeout occurs. A
+// non-positive timeout will block indefinitely.
+func (rb *RingBuffer) Poll(timeout time.Duration) (interface{}, error) {
```

Comment:
I thought so too, but what I ended up doing was:

```go
// Get a byte array from the pool.
func (p *RBPool) Get() []byte {
        if p.ring.Len() == 0 {
                return make([]byte, alloc)
        }
        b, _ := p.ring.Poll(1)
        if b == nil {
                b = make([]byte, alloc)
        }
        return b.([]byte)
}
```
Works OK. `sync.Pool` is much better now though:
```
BenchmarkSlab-8                         20000000                83.5 ns/op             0 B/op          0 allocs/op                                                                    │
BenchmarkSlabParallel-8                 10000000               179 ns/op               0 B/op          0 allocs/op                                                                    │
BenchmarkMake-8                          3000000               547 ns/op            4096 B/op          1 allocs/op                                                                    │
BenchmarkMakeParallel-8                  3000000               520 ns/op            4096 B/op          1 allocs/op                                                                    │
BenchmarkSyncPool-8                     20000000                75.1 ns/op            32 B/op          1 allocs/op                                                                    │
BenchmarkSyncPoolParallel-8             50000000                24.0 ns/op            32 B/op          1 allocs/op                                                                    │
BenchmarkChannelPool-8                  20000000                71.6 ns/op             0 B/op          0 allocs/op                                                                    │
BenchmarkChannelPoolParallel-8           5000000               287 ns/op               0 B/op          0 allocs/op                                                                    │
BenchmarkRBPool-8                       20000000               111 ns/op              32 B/op          1 allocs/op                                                                    │
BenchmarkRBPoolParallel-8               10000000               158 ns/op             890 B/op          1 allocs/op                                                                    │
BenchmarkArena-8                          200000                99.0 ns/op             0 B/op          0 allocs/op                                                                    │
BenchmarkArenaParallel-8                  200000                28.0 ns/op             0 B/op          0 allocs/op
```